src/ralph/backends/data/fs.py

```diff
@@ -0,0 +1,291 @@
+                If target is a relative path, it is considered to be relative to the
+                    `default_directory_path`.
+                If target is `None`, a random (uuid4) file is created in the
+                    `default_directory_path` and used as the target instead.
+            chunk_size (int or None): Ignored. Is equal to the default chunk size.
+            ignore_errors (bool): Ignored.
+            operation_type (BaseOperationType or None): The mode of the write operation.
+                If operation_type is `CREATE` or `INDEX`, the target file is expected to
+                    be absent. If the target file exists a `FileExistsError` is raised.
+                If operation_type is `UPDATE`, the target file is overwritten.
+                If operation_type is `APPEND`, the data is appended to the
+                    end of the target file.
+
+        Returns:
+            int: The number of written files.
+        """
+        data = iter(data)
+        try:
+            first_record = next(data)
+        except StopIteration:
+            logger.info("Data Iterator is empty; skipping write to target.")
+            return 0
+        if not operation_type:
+            operation_type = self.default_operation_type
+
+        if operation_type == BaseOperationType.DELETE:
+            msg = "Delete operation_type is not allowed."
+            logger.error(msg)
+            raise BackendParameterException(msg)
+
+        if not target:
+            target = f"{now()}-{uuid4()}"
+            logger.info("Target file not specified; using random file name: %s", target)
+
+        target = Path(target)
+        path = target if target.is_absolute() else self.default_directory / target
+
+        if operation_type in [BaseOperationType.CREATE, BaseOperationType.INDEX]:
+            if path.is_file():
+                msg = (
+                    "%s already exists and overwrite is not allowed with operation_type"
+                    " create or index."
+                )
+                logger.error(msg, path)
+                raise FileExistsError(msg % path)
+
+            logger.debug("Creating file: %s", path)
+
+        mode = "wb"
+        if operation_type == BaseOperationType.APPEND:
+            mode = "ab"
+            logger.debug("Appending to file: %s", path)
+
+        with path.open(mode) as file:
+            is_dict = isinstance(first_record, dict)
+            writer = self._write_dict if is_dict else self._write_raw
+            writer(file, first_record)
+            for chunk in data:
+                writer(file, chunk)
+
+        # The file has been created, add a new entry to the history.
+        self.append_to_history(
+            {
+                "backend": self.name,
+                "action": "write",
+                "id": str(path),
+                "filename": path.name,
+                "size": path.stat().st_size,
+                "timestamp": now(),
+            }
```

Comment:
Probably outside of this PR scope, but I think it would be nice to normalize history entries, at least for the core, with something looking like this:
``` python
class HistoryEntry(BaseModel):
""" Pydantic model for history entries."""
    backend: str
    action: str
    id: str
    size: int
    timestamp: datetime

    class Config:
        """HistoryEntry Model configuration."""
        extra = "allow"
```
It would avoid mistakes when adding new backends (`get_command_history` requires entry to have `backend` and `command` keys).
WDYT?

Reply:
Good idea!

Reply:
FTR: it will be changed in #297 by @wilbrdt